### PR TITLE
Add flowgraph manager to manage datasyncservices

### PR DESCRIPTION
**Pull request by XuanYang-cn**
- Replace the 2 maps with sync.map
- Add ut to 90%

Signed-off-by: yangxuan <xuan.yang@zilliz.com>

/kind improvement

### CODE REVIEW

**Comment by XuanYang-cn** [comment]
rerun ut

**Comment by XuanYang-cn** [comment]
#14604

File: internal/datanode/flow_graph_manager.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+type flowgraphManager struct {
+	flowgraphs sync.Map // vChannelName -> dataSyncService
+}
+
+func newFlowgraphManager() *flowgraphManager {
+	return &flowgraphManager{}
+}
+
+func (fm *flowgraphManager) AddAndStart(dn *DataNode, vchan *datapb.VchannelInfo) error {
```
Comment (sunby) [inline internal/datanode/flow_graph_manager.go:37]: If this struct is only used in package datanode, we could use lower case to name the function

File: internal/datanode/flow_graph_manager.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	log.Debug("Received Vchannel Info",
```
Comment (sunby) [inline internal/datanode/flow_graph_manager.go:38]: No need to start a log with upper case

File: internal/datanode/flow_graph_manager.go (hunk 1 above)
Comment (sunby) [reply internal/datanode/flow_graph_manager.go:37]: And I suggest to change the name to `add`

File: internal/datanode/flow_graph_manager.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	var alloc allocatorInterface = newAllocator(dn.rootCoord)
+
+	dataSyncService, err := newDataSyncService(dn.ctx, make(chan flushMsg, 100), replica, alloc, dn.msFactory, vchan, dn.clearSignal, dn.dataCoord, dn.segmentCache, dn.blobKv, dn.compactionExecutor)
+	if err != nil {
+		log.Warn("newDataSyncService fail", zap.String("vChannelName", vchan.GetChannelName()), zap.Error(err))
+		return err
+	}
+	log.Info("successfully created dataSyncService", zap.String("vChannelName", vchan.GetChannelName()))
```
Comment (sunby) [inline internal/datanode/flow_graph_manager.go:62]: If the `dataSyncService` is not proper noun, we can just print `data sync service` or ` service to sync data` to make it clear.

File: internal/datanode/data_node.go [hunk 4]
```diff
@@ -306,130 +299,65 @@ func (node *DataNode) handleChannelEvt(evt *clientv3.Event) {
...
-
 // BackGroundGC runs in background to release datanode resources
 func (node *DataNode) BackGroundGC(vChannelCh <-chan string) {
 	log.Info("DataNode Background GC Start")
 	for {
 		select {
 		case vChan := <-vChannelCh:
 			log.Info("GC flowgraph", zap.String("vChan", vChan))
-			node.ReleaseDataSyncService(vChan)
+			go node.flowgraphMg.Release(vChan)
```
[the remark this answers is not in the document]
Comment (XuanYang-cn) [reply internal/datanode/data_node.go:353]: It's guaranteed by  `sync.Map` `LoadAndDelete`

File: internal/datanode/event_manager.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+type eventManager struct {
+	eventChan         chan event
+	closeChan         chan struct{}
+	handlePutEvent    func(watchInfo *datapb.ChannelWatchInfo) error // node.handlePutEvent
+	handleDeleteEvent func(vChanName string)                         // node.handleDeleteEvent
+}
+
+var eventManagerMap = make(map[string]*eventManager)
```
Comment (XuanYang-cn) [inline internal/datanode/event_manager.go:44]: need to delete

File: internal/datanode/data_node.go [hunk 6]
```diff
@@ -287,149 +282,141 @@ func (node *DataNode) checkWatchedList() error {
...
+			eventType: deleteEventType,
+		}
+	}
+	node.handleWatchInfo(e, string(evt.Kv.Key), evt.Kv.Value)
+}
+
+func (node *DataNode) handleWatchInfo(e *event, key string, data []byte) {
+	switch e.eventType {
+	case putEventType:
+		log.Info("DataNode handling watchInfo put event", zap.String("key", key))
```
Comment (sunby) [inline internal/datanode/data_node.go:310]: correct the grammar

File: internal/datanode/event_manager.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+								if err == nil {
+									log.Debug("retry to handle put event successfully",
+										zap.String("vChanName", event.vChanName))
+									return
+								}
+							}
+						}
+					}()
+
+					wg.Wait()
```
Comment (congqixia) [inline internal/datanode/event_manager.go:94]: Is this goroutine needed?

File: internal/datanode/event_manager.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+const retryWatchInterval = 20 * time.Second
+
+type event struct {
+	eventType int
+	vChanName string
+	info      *datapb.ChannelWatchInfo
+}
+
+type eventManager struct {
```
Comment (sunby) [inline internal/datanode/event_manager.go:37]: Is it better to change the name to `channelEventManager`?

File: internal/datanode/flow_graph_manager.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		log.Warn("new data sync service fail", zap.String("vChannelName", vchan.GetChannelName()), zap.Error(err))
+		return err
+	}
+	log.Info("successfully created dataSyncService", zap.String("vChannelName", vchan.GetChannelName()))
+
+	dataSyncService.start()
+	log.Info("successfully started dataSyncService", zap.String("vChannelName", vchan.GetChannelName()))
+
+	fm.flowgraphs.Store(vchan.GetChannelName(), dataSyncService)
+	return nil
```
Comment (congqixia) [inline internal/datanode/flow_graph_manager.go:68]: This function may leak datasyncservice if two or more simultaneous call comes

File: internal/datanode/event_manager.go (hunk 7 above)
Comment (XuanYang-cn) [reply internal/datanode/event_manager.go:94]: no

File: internal/datanode/flow_graph_manager.go (hunk 9 above)
Comment (XuanYang-cn) [reply internal/datanode/flow_graph_manager.go:68]: event manager will make the watch queuing for the same channel

File: internal/datanode/event_manager.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+	handlePutEvent    func(watchInfo *datapb.ChannelWatchInfo) error // node.handlePutEvent
+	handleDeleteEvent func(vChanName string)                         // node.handleDeleteEvent
+}
+
+const (
+	putEventType    = 1
+	deleteEventType = 2
+)
+
+func (e *channelEventManager) Run() {
```
Comment (czs007) [inline internal/datanode/event_manager.go:48]: better to add sync.waitGroup and wait in Close().
better to extract the main logic into a separate function.

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**